File: 2.linked-list/single-linked-list/reverse-linked-list/reverse.py
# Iterative method

def reverse(self, head):
  prev = None
  current = head
  while current:
    next = current.next
    current.next = prev
    prev = current
    current = next
  return prev
  
# Recursive method
def reverse2(self, head):
  return self._reverse(head)

def _reverse(self, node, prev=None):
  if not node:
    return None
  next = node.next
  node.next = prev
  return self._reverse(next,node)

File: 2.linked-list/single-linked-list/reverse-linked-list/reverse.js
// Iterative method
let reverse = function (head) {
  let prev = null;
  let current = head;
  let next = null;
  while (current) {
    next = current.next;
    current.next = prev;
    prev = current;
    current = next;
  }
  return prev;
};

// Recursive method
let reverse2 = function (head) {
  return _reverse(head);
};

let _reverse = function (node, prev = null) {
  if (!head) {
    return null;
  }
  const next = node.next;
  node.next = prev;
  return _reverse(next, node);
};
